provider/cmd/pulumi-resource-pulumiservice/schema.json

```diff
@@ -1219,6 +1219,47 @@
         "stack",
         "timestamp"
       ]
+    },
+    "pulumiservice:index:Environment": {
+      "description": "An ESC Environment.",
+      "properties": {
+        "organization": {
+          "description": "Organization name.",
+          "type": "string"
+        },
+        "name": {
+          "description": "Environment name.",
+          "type": "string"
+        },
+        "yaml": {
+          "description": "Environment's yaml file.",
+          "$ref": "pulumi.json#/Asset"
+        }
+      },
+      "required": [
+        "organization",
+        "name",
```

Comment:
In other providers we have a pattern of allowing the name to be optional and pulling it from the URN during `Check` if it's unspecified. I think it's okay to leave it required here--I think we can make it optional + compute it in Check in the future w/o a breaking change if we choose.

Reply:
Do I understand it right that the name is not stored separately as a property and just pulled from id every time? I'm not sure I understand the benefit apart from saving a bit of DB space... Would you link an example please?

Reply:
It's a UX benefit--it keeps users from repeating names and allows the provider to add entropy in order to avoid collisions between multiple deployments of the same Pulumi program.

For example, the following code will create a bucket named `foo-<some random suffix>`:

```typescript
const b = new aws.s3.Bucket("foo");
```

But this code will create a bucket named `foo`:

```typescript
const b = new aws.s3.Bucket("foo", { bucket: "foo" });
```

In the first example, the provider's `Check` method will pick `foo` out of the "name" portion of the resource's URN, attach some entropy to the end, and inject the final name into the result (unless there is a name present in the state, in which case it will use that name).

I'm not sure that this is behavior we _want_ for the `Environment` resource, but it's something that's worth considering.

Reply:
Ack, makes sense, thank you for the laid out example

I think it might look ugly if we by default have `EnvironmentName-abc123` in the console after customer named the resource `EnvironmentName`. For things like s3 buckets you probably don't care what the name is, for environment I think people would care more. But also if customers are creating environments via provider, they aren't likely to go look at it in the console, so not sure.

Let me know if you'd prefer I add this logic, I don't know enough about providers yet to have a strong opinion

Reply:
FWIW @pgavlin we don't have autonaming for any resources in this provider yet. I think let's leave it as-is and address auto-naming wholesale?

Reply:
SGTM